WordPress/gutenberg
# `Tabs`: tweak sizing and overflow behavior of TabList

## What?


This PR modifies the behavior of `Tabs.TabList` in regards to sizing and overflow to address some needs.

## Why?


We have needs around proper management of overflowing tabs (https://github.com/WordPress/gutenberg/issues/64093) and tab lists that expand to fill their container (https://github.com/WordPress/gutenberg/pull/61974#issuecomment-2271217124).

## How?


- It makes the `TabList` have a default `width` of `fit-content` (only for horizontal orientation). This means that tabs will only take up the space needed for their content, and the tab list (container) will adjust its width to the space needed for the sum of its child tabs.
  - This is set with `:where` for reduced specificity, in order to make overrides easier.
- It sets `overflow-x` to `auto` in the `TabList`, making it overflow horizontally only when necessary.
- It makes tabs expand (`flex-grow: 1`) to fill up **additional** available space when the container (`TabList`) has it. Note that due to `TabList` having `width: fit-content` by default, this won't happen unless the user intentionally modifies the width, for example by setting `width: 100%` to fill up its container's size.
- It centers the contents of a tab horizontally (`justify-content: center`, only for horizontal orientation).

The TL;DR is:

- Tabs will scroll when necessary.
- Tabs hug contents by default.
- If a "fill" style is wanted, just set the `TabList` width you want and tabs will adapt gracefully.

There is an additional detail that is important now that the tab list scrolls: when a tab is selected, it is programmatically scrolled into view.

Finally, I've gone through every single instance of `Tabs` and applied any necessary tweaks to retain the same styles. Everything should look exactly the same after these changes.

## Testing Instructions





Open the storybook locally (`npm run storybook:dev`) and go to the "Tabs > Size And Overflow Playground" story. Play with it. Also, check that none of the instances of this component in Gutenberg are affected (see below).

---

Where to find all `Tabs` instances for testing:

1. This one requires a small tweak in the code, replace `canChooseAColor && canChooseAGradient` with `true` in `gutenberg/packages/block-editor/src/components/colors-gradients/control.js`, then open post editor > select some text > go to "Block" in right sidebar > click "Background" (it'll be the "second" tablist below the first one)
2. Open post editor > select some text > go to "Block" in right sidebar > click "Background"
3. Open post editor > open block inserter > check inside "Patterns" and "Media" (vertical tabs)
4. Open post editor > add a "navigation block" and select it > go to "Block" in the right sidebar (tabs with icon labels)
5. Open post editor > open block inserter and document overview
6. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Select a palette
7. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Typography > Manage fonts
8. (With a "block" theme set like Twenty Twenty-Four) Appearance > Editor > Styles > Edit styles (pencil icon button) > Style Book (eye icon button)
9. (With a "non-block" theme set like Twenty Fifteen) Appearance > Widgets (tabs i right sidebar)
10. Open post editor > open right sidebar
11. Open post editor > select some text > open "more" dropdown in the floating menu > Highlight
12. Open post editor > open options dropdown menu (vertical dots icon button) > Preferences (vertical tabs)

### Testing Instructions for Keyboard


In the same story, you can tab into the tab list and use arrows to move.

## Screenshots or screencast 

https://github.com/user-attachments/assets/ac2bdac8-bc7a-437e-8b62-de5dd3d1360b
